src/java/org/apache/cassandra/index/sai/utils/PrimaryKeys.java

```diff
@@ -33,29 +33,41 @@ public class PrimaryKeys implements Iterable<PrimaryKey>
     private static final long EMPTY_SIZE = ObjectSizes.measure(new PrimaryKeys());
 
     // from https://github.com/gaul/java-collection-overhead
-    private static final long SET_ENTRY_OVERHEAD = 36;
+    private static final long MAP_ENTRY_OVERHEAD = 36;
 
-    private final ConcurrentSkipListSet<PrimaryKey> keys = new ConcurrentSkipListSet<>();
+    // We store a mapping from PrimaryKey to an Object to allow for put-first semantics with a subsequent remove
+    // that only removes the key if the object reference is different from the one that was put. This is used
+    // to simplify updates to collection columns and to analyzed text columns.
```

Comment:
I added a new commit with that uses the set based solution. I think it would be interesting to consider another solution that finds the `net new` and `net old` terms for removal, but this issue https://github.com/riptano/cndb/issues/10584 with analyzed `list<text>` columns complicates the design, so I haven't switched yet.

I think it would be helpful to benchmark the difference in designs to see what each costs. I'm not sure I have time/priority to do that though.

Reply:
I will update the PR description once we settle on a design.